ff_control/scripts/opt_ctrl_py_node

```diff
@@ -0,0 +1,505 @@
+            # Add equality dynamics constraint 
+            g   += [Xk_end-Xk]
+            lbg += [0, 0, 0, 0, 0, 0]
+            ubg += [0, 0, 0, 0, 0, 0]
+
+        J = J + 10*self.k_pos*(self.normsq(Xk_end[0:2]-goal[0:2])) + 10*self.k_th*(1-cos(goal[2]-Xk_end[2])) + 10*self.k_velo*self.normsq(Xk_end[3:]-goal[3:])
+
+        # Concatenate decision variables and constraint terms
+        w = vertcat(*w)
+        g = vertcat(*g)
+
+        # Create an NLP solver
+        nlp_prob = {'f': J, 'p':goal, 'x': w, 'g': g}
+        cont_nlp_solver = nlpsol('nlp_solver', 'ipopt', nlp_prob); # Solve relaxed problem
+        return w0, cont_nlp_solver, lbw, ubw, lbg, ubg
+
+    ############################### Helper Functions for Opt ###############################
+    def _map_to_force(self, u):
+        # Compute body-frame force from thrusters
+        Fx = -u[0] + u[2] 
+        Fy = -u[1] + u[3] 
+        M = self.r * (u[0]+u[1]+u[2]+u[3])
+        return Fx, Fy, M
+
+    def normsq(self, x):
+        sum = 0
+        for i in range(x.shape[0]):
+            sum += x[i]**2
+        return sum
+
+    def normabs(self, x):
+        sum = 0
+        for i in range(x.shape[0]):
+            sum += fabs(x[i])
+        return sum
+
+    def unpack_wopt(self, w_opt):
+        w_opt = w_opt.full().flatten()
+        u0_opt = w_opt[6::10]
+        u1_opt = w_opt[7::10]
+        u2_opt = w_opt[8::10]
+        u3_opt = w_opt[9::10]
+        return u0_opt, u1_opt, u2_opt, u3_opt
+
+    def get_next_warm_start(self, w_opt):
+        output = w_opt.full().flatten()     
+        x = [output[i:6+i] for i in range(w_opt.size()[0]//10+1)]
+        u = [output[6+i:10+i] for i in range(w_opt.size()[0]//10)]
+
+        w0 = [DM(x[1])]
+        for i in range(len(u)-1):
+            w0 += [DM(u[i+1])]
+            w0 += [DM(x[i+2])]
+        w0 += [DM(u[-1])]
+        w0 += [DM(x[-1])]
+
+        return w0
+
+    ############################### Helper Functions to unpack FreeFlyerState ###############################
+    @staticmethod
+    def state2vec(state: FreeFlyerState) -> np.ndarray:
+        """
+        Convert state message to state vector.
+
+        :param state: state message
+        :return: state vector
+        """
+        return np.array(
+            [
+                state.pose.x,
+                state.pose.y,
+                state.pose.theta,
+                state.twist.vx,
+                state.twist.vy,
+                state.twist.wz,
+            ]
+        )
+
+    @staticmethod
+    def vec2state(vec: np.ndarray) -> FreeFlyerState:
+        """
+        Convert state vector to state message.
+
+        :param vec: state vector
+        :return: state message
+        """
+        state = FreeFlyerState()
+        state.pose.x = vec[0]
+        state.pose.y = vec[1]
+        state.pose.theta = vec[2]
+        state.twist.vx = vec[3]
+        state.twist.vy = vec[4]
+        state.twist.wz = vec[5]
+
+        return state
```

Comment:
Is it copied from some other node? Maybe we can create a common utility function for this instead of duplicating code?